### PR TITLE
Check function for null.

### PR DESCRIPTION
**Pull request by SchoolGuy**
## What does this PR change?

Should fix bsc#1136286

## GUI diff

No difference.

- [X] **DONE**

## Documentation
- No documentation needed: Bugfix in the internal code

- [X] **DONE**

## Test coverage
- No tests: Tests are existing (this is a fix for a failing one).

- [X] **DONE**

## Links

Fixes bsc#1136286
Tracks https://github.com/SUSE/spacewalk/issues/8002

- [X] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [X] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"		 
- [ ] Re-run test "java_pgsql_tests"		 
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"		 
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"		 
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by SchoolGuy** [comment]
@lneves12 @ncounter Would one (or both of you) review this pls?

File: java/code/src/com/suse/manager/utils/SaltUtils.java [hunk 1]
```diff
@@ -251,7 +251,7 @@ public static PackageChangeOutcome handlePackageChanges(String function,
             ErrataManager.insertErrataCacheTask(server);
             outcome = applyChangesFromStateModule(delta, server);
         }
-        else if (function.equals("state.apply")) {
+        else if (function != null && function.equals("state.apply")) {
```
Comment (ncounter) [inline java/code/src/com/suse/manager/utils/SaltUtils.java]: Did you had the chance to test this and see that it does not fail anymore?

Anyway, since the `String.equals` allows receiving `null` arguments, why not just reverting the order of the check instead of adding another one? `else if ("state.apply".equals(function))`

Comment (SchoolGuy) [reply java/code/src/com/suse/manager/utils/SaltUtils.java]: @ncounter I will deploy my branch to refhead and see if it works

Comment (SchoolGuy) [reply java/code/src/com/suse/manager/utils/SaltUtils.java]: The fix works but now there is another error. It expects the wrong text. Is this another PR or shall I fix that also?

Comment (ncounter) [reply java/code/src/com/suse/manager/utils/SaltUtils.java]: If the error is different and the line affected is different, it's definitely a different PR, even a different bug I'd say. What is failing and how?

Comment (ncounter) [reply java/code/src/com/suse/manager/utils/SaltUtils.java]: At least, the fix deserve a separate commit with the proper bugnumber in the commit text. If you want to add the second fix here, let's do it but keep it in different commits.

Comment (SchoolGuy) [reply java/code/src/com/suse/manager/utils/SaltUtils.java]: Okay then I will investigate deeper and decide then if this error deserves an own PR or if I just do a commit.

**Comment by SchoolGuy** [comment]
Bug https://bugzilla.suse.com/show_bug.cgi?id=1142369 is possibly related to this one.

File: java/code/src/com/suse/manager/utils/SaltUtils.java (hunk 1 above)
Comment (lucidd) [reply java/code/src/com/suse/manager/utils/SaltUtils.java]: why is `function` null here? The method clearly expects a String and passing anything else to it is invalid use. There is a lot of other functions in this file and probably around the whole codebase taking in the exact same `function` variable expecting it to be a String. This is not solved by adding a null check here but by finding out where the null comes from and properly marking the source as `Optional`. This way all places directly or indirectly getting function from this source will become compile errors until the `Optional` is handled.

Comment (ncounter) [reply java/code/src/com/suse/manager/utils/SaltUtils.java]: @SchoolGuy I remember you told me you based this fix on the assumption that it is fine to have a `null` function value here (did you discussed it with some Salt guy?). Maybe this assumption needs to be revisited/dropped according to @lucidd 's comment which is definitely correct. If the `function` key is evaluated in many places and expected to be `String`, we definitely need something different than this check.

Comment (SchoolGuy) [reply java/code/src/com/suse/manager/utils/SaltUtils.java]: @ncounter @lucidd Then I will investigate deeper and try to find the real cause.

**Comment by mcalmer** [comment]
@SchoolGuy can you please summarize the result of your research so somebody can take over?

**Comment by SchoolGuy** [comment]
I think I discussed this after a standup with @ncounter or @lneves12 a while ago. It seems that some salt event cannot be parsed correctly and thus this variable is then null. I was not able to trace it back to a specific one and I was not able to confirm if this is the root cause of this problem. Sorry to not be able to give you more information.

**Comment by ncounter** [comment]
@mcalmer this PR was a quick fix we decided better to not merge before understanding what is really going on and why at the Salt Return Event that in some case returns data without the **expected** `fun` value. As per we expect that `key:value` in many other places, it would be better to figure out what should really be the expectation from the event and handling the data properly in the code path. It is ongoing within [this bug](https://bugzilla.suse.com/show_bug.cgi?id=1142369#c31) where others are flagged as duplicated and are related and more specific against this issue. I remember I already handed over those info to @cbbayburt and I guess he handed over to @mbologna who is currently assigned to the bug.

Bottom line: this PR should be closed now. The proper handling will be implemented according to the bug report. Anything against?

**Comment by mbologna** [comment]
The bug in question was reported by two customers:
- one was having problem with tomcat, resolved by adding localhost to /etc/hosts (configuration problem)
- the other one was having problems with salt-api and /run collection results with a particular host.

**Comment by SchoolGuy** [comment]
Since @ncounter said this should be closed, I will do so now.

**Comment by mcalmer** [comment]
@ncounter the bug you referenced does not really discuss the missing `fun` as it reference a patch which should add a user to the notification websocket. So I am a bit skeptical that somebody is researching why this happens.
I hope we remember this and can talk about it when you are back.
